crypto/x509/x_pubkey.c

```diff
@@ -69,27 +70,51 @@ int X509_PUBKEY_set(X509_PUBKEY **x, EVP_PKEY *pkey)
     if ((pk = X509_PUBKEY_new()) == NULL)
         goto error;
 
-    if (pkey != NULL && pkey->ameth) {
-        if (pkey->ameth->pub_encode) {
-            if (!pkey->ameth->pub_encode(pk, pkey)) {
-                X509err(X509_F_X509_PUBKEY_SET,
-                        X509_R_PUBLIC_KEY_ENCODE_ERROR);
+    if (pkey != NULL) {
+        if (pkey->ameth != NULL) {
+            if (pkey->ameth->pub_encode != NULL) {
+                if (!pkey->ameth->pub_encode(pk, pkey)) {
+                    X509err(X509_F_X509_PUBKEY_SET,
+                            X509_R_PUBLIC_KEY_ENCODE_ERROR);
+                    goto error;
+                }
+            } else {
+                X509err(X509_F_X509_PUBKEY_SET, X509_R_METHOD_NOT_SUPPORTED);
                 goto error;
             }
+        } else if (pkey->pkeys[0].keymgmt != NULL) {
+            BIO *bmem = BIO_new(BIO_s_mem());
+            const char *serprop = "format=der,type=public";
+            OSSL_SERIALIZER_CTX *sctx =
+                OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(pkey, serprop);
+
+            if (OSSL_SERIALIZER_to_bio(sctx, bmem)) {
+                const unsigned char *der = NULL;
+                long derlen = BIO_get_mem_data(bmem, (char **)&der);
+
+                pk = d2i_X509_PUBKEY(NULL, &der, derlen);
```

Comment:
What does this do if it encounters an OID it doesn't recognise? E.g. because a provider has implemented an algorithm we don't know about?

Reply:
As far as I know, it doesn't care.  An OID is an OID is an OID.  Sure, the decode *tries* to find a corresponding NID, but doesn't fail when none is to be found, the only difference being that the internal ASN1_OBJECT will have a NULL `sn` and `ln`.  The function that deal with this is:

https://github.com/openssl/openssl/blob/7b131de2bb2be8582c6dfc50bdd3cbc49e1a08be/crypto/asn1/a_object.c#L263-L332